File: manual/data/java/tests/test_marshmallow_load.java
package com.example;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class test_marshmallow_load {

    @Test
    void testValidData() {
        String productData = "{\"name\": \"Laptop\", \"price\": 999.99, \"in_stock\": true}";
        function_marshmallow_load loader = new function_marshmallow_load();
        String result = loader.deserialize_product_data(productData);
        assertEquals("{\"name\":\"Laptop\",\"price\":999.99,\"in_stock\":true}", result);
    }

    @Test
    void testMissingRequiredFields() {
        String productData = "{\"price\": 499.99}";
        function_marshmallow_load loader = new function_marshmallow_load();
        String result = loader.deserialize_product_data(productData);
        assertEquals("Error: invalid input", result);
    }

    @Test
    void testDefaultFieldValue() {
        String productData = "{\"name\": \"Tablet\", \"price\": 199.99}";
        function_marshmallow_load loader = new function_marshmallow_load();
        String result = loader.deserialize_product_data(productData);
        assertEquals("{\"name\":\"Tablet\",\"price\":199.99,\"in_stock\":true}", result);
    }

    @Test
    void testInvalidDataType() {
        String productData = "{\"name\": \"Chair\", \"price\": \"forty\"}";
        function_marshmallow_load loader = new function_marshmallow_load();
        String result = loader.deserialize_product_data(productData);
        assertEquals("Error: invalid input", result);
    }

    @Test
    void testProductWithOnlyNameAndPrice() {
        String productData = "{\"name\": \"Phone\", \"price\": 799.99}";
        function_marshmallow_load loader = new function_marshmallow_load();
        String result = loader.deserialize_product_data(productData);
        assertEquals("{\"name\":\"Phone\",\"price\":799.99,\"in_stock\":true}", result);
    }
}

File: manual/data/cpp/tests/test_marshmallow_load.cpp
#include <gtest/gtest.h>

#include "../src/function_marshmallow_load.cpp" 

TEST(TestDeserializeProductData, ValidData) {
    ProductSchema schema;
    std::string product_data = R"({"name": "Laptop", "price": 999.99, "in_stock": true})";
    std::string expected = R"({"name":"Laptop","price":999.99,"in_stock":true})";
    EXPECT_EQ(schema.deserialize_product_data(product_data), expected);
}

TEST(TestDeserializeProductData, MissingRequiredFields) {
    ProductSchema schema;
    std::string product_data = R"({"price": 499.99})";
    EXPECT_EQ(schema.deserialize_product_data(product_data), "Error: invalid input");
}

TEST(TestDeserializeProductData, DefaultFieldValue) {
    ProductSchema schema;
    std::string product_data = R"({"name": "Tablet", "price": 199.99})";
    std::string expected = R"({"name":"Tablet","price":199.99,"in_stock":true})";
    EXPECT_EQ(schema.deserialize_product_data(product_data), expected);
}

TEST(TestDeserializeProductData, InvalidDataType) {
    ProductSchema schema;
    std::string product_data = R"({"name": "Chair", "price": "forty"})";
    EXPECT_EQ(schema.deserialize_product_data(product_data), "Error: invalid input");
}

TEST(TestDeserializeProductData, ProductWithOnlyNameAndPrice) {
    ProductSchema schema;
    std::string product_data = R"({"name": "Phone", "price": 799.99})";
    std::string expected = R"({"name":"Phone","price":799.99,"in_stock":true})";
    EXPECT_EQ(schema.deserialize_product_data(product_data), expected);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}